File: FreestarOBJCSample/FreestarOBJCSample/PartnerSelectionTableViewController.m
//
//  PartnerSelectionTableViewController.m
//
//  Created by Lev Trubov on 10/8/19.
//  Copyright © 2019 Lev Trubov. All rights reserved.
//

#import "PartnerSelectionTableViewController.h"
#import "PartnerChoice.h"

@interface PartnerSelectionTableViewController () {
    NSArray<PartnerChoice *> *partnerList;
}

@property(weak) id<PartnerSelectionViewControllerDelegate> delegate;

@end

@implementation PartnerSelectionTableViewController

-(instancetype)initWithAdType:(NSString *)adType andDelegate:(id<PartnerSelectionViewControllerDelegate>)delegate {
    self = [super init];
    partnerList = [self partnerListForAdType:adType];
    self.delegate = delegate;
    return self;
}

-(NSArray<PartnerChoice *> *)partnerListForAdType:(NSString *)adType {
    if([adType isEqualToString:@"Interstitial"]) {
        return @[
            [[PartnerChoice alloc] initWithName:@"All" andSelected:YES],
            [[PartnerChoice alloc] initWithName:@"AdColony"],
            [[PartnerChoice alloc] initWithName:@"GoogleAdmob"],
            [[PartnerChoice alloc] initWithName:@"Unity"],
            [[PartnerChoice alloc] initWithName:@"AppLovin"],
            [[PartnerChoice alloc] initWithName:@"Vungle"],
            [[PartnerChoice alloc] initWithName:@"Criteo"],
            [[PartnerChoice alloc] initWithName:@"Google"],
            [[PartnerChoice alloc] initWithName:@"Nimbus"],
            [[PartnerChoice alloc] initWithName:@"TAM"]
        ];
    } else if([adType isEqualToString:@"Rewarded"]) {
        return @[
            [[PartnerChoice alloc] initWithName:@"All" andSelected:YES],
            [[PartnerChoice alloc] initWithName:@"Auction"],
            [[PartnerChoice alloc] initWithName:@"Vungle"],
            [[PartnerChoice alloc] initWithName:@"AdColony"],
            [[PartnerChoice alloc] initWithName:@"GoogleAdmob"],
            [[PartnerChoice alloc] initWithName:@"Unity"],
            [[PartnerChoice alloc] initWithName:@"AppLovin"],
            [[PartnerChoice alloc] initWithName:@"Tapjoy"],
            [[PartnerChoice alloc] initWithName:@"Criteo"],
            [[PartnerChoice alloc] initWithName:@"Nimbus"]
        ];
    } else if([adType isEqualToString:@"Banner"]) {
        return @[
            [[PartnerChoice alloc] initWithName:@"All" andSelected:YES],
            [[PartnerChoice alloc] initWithName:@"AdColony"],
            [[PartnerChoice alloc] initWithName:@"AppLovin"],
            [[PartnerChoice alloc] initWithName:@"GoogleAdmob"],
            [[PartnerChoice alloc] initWithName:@"Criteo"],
            [[PartnerChoice alloc] initWithName:@"Google"],
            [[PartnerChoice alloc] initWithName:@"Nimbus"],
            [[PartnerChoice alloc] initWithName:@"TAM"]
        ];
    } else if([adType isEqualToString:@"Preroll"]) {
        return @[
            [[PartnerChoice alloc] initWithName:@"All" andSelected:YES],
            [[PartnerChoice alloc] initWithName:@"Google"],
            [[PartnerChoice alloc] initWithName:@"Nimbus"]
        ];
    } else if([adType isEqualToString:@"Small Banner"]) {
        return @[
            [[PartnerChoice alloc] initWithName:@"All" andSelected:YES],
            [[PartnerChoice alloc] initWithName:@"GoogleAdmob"],
            [[PartnerChoice alloc] initWithName:@"Criteo"],
            [[PartnerChoice alloc] initWithName:@"AppLovin"],
            [[PartnerChoice alloc] initWithName:@"Unity"],
            [[PartnerChoice alloc] initWithName:@"Google"],
            [[PartnerChoice alloc] initWithName:@"Nimbus"],
            [[PartnerChoice alloc] initWithName:@"TAM"]

        ];
    } else if([adType isEqualToString:@"Medium Native"]) {
        return @[
            [[PartnerChoice alloc] initWithName:@"All" andSelected:YES]
        ];
    } else if([adType isEqualToString:@"Small Native"]) {
        return @[
            [[PartnerChoice alloc] initWithName:@"All" andSelected:YES]
        ];
    }
    
    return @[]; //shouldn't happen
}


- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"Select Ad Partner";
    [self.tableView registerClass:UITableViewCell.class forCellReuseIdentifier:@"partnerCell"];
    
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(closeAndExit)];
}

-(void)closeAndExit {
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return partnerList.count;
}

//*
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"partnerCell" forIndexPath:indexPath];
    //
    PartnerChoice *pc = partnerList[indexPath.row];
    cell.textLabel.text = pc.name;
    cell.accessoryType = pc.selected ? UITableViewCellAccessoryCheckmark : UITableViewCellAccessoryNone;
    
    return cell;
}
//*/

#pragma mark - UITableViewDelegate

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    PartnerChoice *pc = partnerList[indexPath.row];
    if([pc.name isEqualToString:@"All"]) {
        for(int i = 1; i < partnerList.count; i++) {
            partnerList[i].selected = NO;
        }
        pc.selected = YES;
    } else {
        partnerList[0].selected = NO;
        pc.selected = !pc.selected;
    }
    
    for(int i = 0; i < [tableView numberOfRowsInSection:0]; i++){
        UITableViewCell *cell = [tableView cellForRowAtIndexPath:[NSIndexPath indexPathForRow:i inSection:0]];
        cell.accessoryType = partnerList[i].selected ? UITableViewCellAccessoryCheckmark : UITableViewCellAccessoryNone;
    }

    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

#pragma mark - sending selected partner info on exit

-(void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    if(self.isMovingFromParentViewController) {
        [self.delegate partnersSelected:[self extractChosenPartners]];
    }
}

-(NSArray *)extractChosenPartners {
    if(partnerList[0].selected) {
        return @[@"all"];
    }
    
    NSMutableArray *res = [[NSMutableArray alloc] init];
    [partnerList enumerateObjectsUsingBlock:^(PartnerChoice * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if(obj.selected) {
            [res addObject:[obj.name lowercaseString]];
        }
    }];
    return res;
}

@end

File: FreestarSwiftSample/FreestarSwiftSample/PartnerSelectionTableViewController.swift
//
//  PartnerSelectionTableViewController.swift
//  FreestarSwiftSample
//
//  Copyright © 2020 Freestar. All rights reserved.
//

import UIKit

struct PartnerChoice {
    var name: String
    var selected = false
}

protocol PartnerSelectionViewControllerDelegate: class {
    func partnersSelected(_ partners: [String])
}

class PartnerSelectionTableViewController: UITableViewController {
    
    var partnerList = [PartnerChoice]()
    weak var delegate: PartnerSelectionViewControllerDelegate?
    
    convenience init(adType: FreestarAdType, delegate: PartnerSelectionViewControllerDelegate? = nil) {
        self.init()
        self.partnerList = partnerListForAdType(adType)
        self.delegate = delegate
    }
    
    func partnerListForAdType(_ type: FreestarAdType) -> [PartnerChoice] {
        switch type {
        case .None: return []
        case .Interstitial: return [PartnerChoice(name: "All", selected: true)] + [
                "Chocolate",
                "AdColony",
                "GoogleAdmob",
                "Unity",
                "AppLovin",
                "Vungle",
                "Amazon",
                "Criteo",
                "Google" ].map { PartnerChoice(name: $0) }
        case .Rewarded: return [PartnerChoice(name: "All", selected: true)] + [
                "Chocolate",
                "AdColony",
                "GoogleAdmob",
                "Unity",
                "AppLovin",
                "Vungle",
                "Amazon",
                "Criteo",
                "Google",
                "Tapjoy" ].map { PartnerChoice(name: $0) }
        case .LargeBanner: return [PartnerChoice(name: "All", selected: true)] + [
                "Chocolate",
                "AppLovin",
                "GoogleAdmob",
                "Amazon",
                "Criteo",
                "Google" ].map { PartnerChoice(name: $0) }
        case .SmallBanner: return [PartnerChoice(name: "All", selected: true)] + [
                "GoogleAdmob",
                "Amazon",
                "Criteo",
                "AppLovin",
                "Unity" ].map { PartnerChoice(name: $0) }
        case .Preroll: return [PartnerChoice(name: "All", selected: true)] + [
                "Chocolate",
                "Google",
                "Amazon" ].map { PartnerChoice(name: $0) }
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        self.title = "Select Ad Partner";
        self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: "partnerCell")
        
        self.navigationItem.rightBarButtonItem =
            UIBarButtonItem(barButtonSystemItem: .done,
                            target: self,
                            action: #selector(PartnerSelectionTableViewController.closeAndExit))
    }
    
    @objc func closeAndExit() {
        self.navigationController?.popViewController(animated: true)
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return partnerList.count
    }

    //*
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "partnerCell", for: indexPath)

        // Configure the cell...
        let pc = partnerList[indexPath.row]
        cell.textLabel?.text = pc.name
        cell.accessoryType = pc.selected ? .checkmark : .none

        return cell
    }
    //*/

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if partnerList[indexPath.row].name == "All"  {
            for i in 1..<partnerList.count {
                partnerList[i].selected = false
            }
            partnerList[indexPath.row].selected = true
        } else {
            partnerList[0].selected = false
            partnerList[indexPath.row].selected.toggle()
        }
        
        for i in 0..<tableView.numberOfRows(inSection: 0) {
            let cell = tableView.cellForRow(at: IndexPath(row: i, section: 0))
            cell?.accessoryType = partnerList[i].selected ? .checkmark : .none
        }
        
        tableView.deselectRow(at: indexPath, animated: true)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        if self.isMovingFromParent {
            self.delegate?.partnersSelected(self.extractChosenPartners())
        }
    }
    
    func extractChosenPartners() -> [String] {
        if partnerList[0].selected {
            return ["all"]
        }
        
        return partnerList.filter { $0.selected }.map { $0.name.lowercased() }
    }
}
